stubs/pywin32/win32com/client/dynamic.pyi

```diff
@@ -1,14 +1,72 @@
 from _typeshed import Incomplete
-from typing import Any
+from typing import Any, Protocol, TypeVar, overload
+from typing_extensions import TypeAlias
+
+import _win32typing
+from win32.lib.pywintypes import IIDType
+from win32com.client import build as build
+
+class _DispatchCreateClass(Protocol):
+    def __init__(
+        self,
+        IDispatch: str | PyIDispatchType | _GoodDispatchTypes | PyIUnknownType,
+        olerepr: build.DispatchItem | build.LazyDispatchItem,
+        userName: str | None = ...,
+        UnicodeToString: None = ...,
+        lazydata: Incomplete | None = ...,
+    ): ...
```

Comment:
Unfortunately, I don't think this works the way you want it to :( `__init__` methods in protocols are an edge case that are kind of weird and somewhat confusing, both at runtime and in how they're treated by static type checkers. The runtime behaviour is a whole other discussion, so let's stick to the static aspects here ;)

Type checkers _don't_ regard `__init__` methods as part of the interface that a user declares when inheriting from `typing.Protocol`. That means that [they won't emit an error](https://mypy-play.net/?mypy=latest&python=3.11&gist=f4d508333db7b95d0c5ade55c32662e1) on code like the following, which you might expect them to flag:

```python
from typing import Protocol

class Foo(Protocol):
    def __init__(self, arg: str) -> None: ...

class Bar:
    def __init__(self, arg1: int, arg2: bytes) -> None: ...

def expects_foo(x: Foo) -> None:
    pass

expects_foo(Bar(42, b"bar"))
```

The reasons why they won't emit an error on code like that has to do with the Liskov Substitution Principle, which is generally considered only to apply to _objects that have already been constructed_ -- the _constructor methods_ of objects are considered outside of the purview of the LSP. Some further reading on this, if you're interested:
- https://stackoverflow.com/questions/69394820/mypy-accepts-an-incompatible-type-in-init-override
- https://softwareengineering.stackexchange.com/questions/431549/the-liskov-substitution-principle-and-python

Reply:
![image](https://user-images.githubusercontent.com/1350584/205730283-549b3029-0534-4f1a-93f4-1d621ce44433.png)
![image](https://user-images.githubusercontent.com/1350584/205730288-693fd81a-d25c-4590-b90a-d179e9633d8d.png)
![image](https://user-images.githubusercontent.com/1350584/205730294-4245bea0-e40c-4216-9049-6539ff4f3062.png)


But I also tried reimplementing it using Callback Protocols and it seems to be working just as well:
```py
_T_co = TypeVar("_T_co", covariant=True)
_T = TypeVar("_T")

class _DispatchCreateClass(Protocol[_T_co]):
    @staticmethod
    def __call__(
        IDispatch: str | PyIDispatchType | _GoodDispatchTypes | PyIUnknownType,
        olerepr: build.DispatchItem | build.LazyDispatchItem,
        userName: str | None = ...,
        UnicodeToString: None = ...,
        lazydata: Incomplete | None = ...,
    ) -> _T_co: ...

@overload
def Dispatch(
    IDispatch: str | PyIDispatchType | _GoodDispatchTypes | PyIUnknownType,
    userName: str | None,
    createClass: _DispatchCreateClass[_T],
    typeinfo: _win32typing.PyITypeInfo | None = ...,
    UnicodeToString: None = ...,
    clsctx: int = ...,
) -> _T: ...
```

Sidenote, the parameter `createClass` of the first overload should not be ` | None` either way. Fixing that as well.

Reply:
If I change the *type* of a parameter, mypy won't catch it with `__init__`. So Callback protocols it is.

Fun fact, pyright did catch it however:
![image](https://user-images.githubusercontent.com/1350584/205732075-aab6acd8-989c-43e8-8545-c69a6a81dfbc.png)